### DISCUSSION TITLE
FIRST() and LAST() - where are the docs?

### DISCUSSION THREAD

**Discussion by dukester** [modula3/cm3#808] [Q&A]
Can't find them nowhere at:
[https://modula3.github.io/cm3/tutorial/m3/m3_toc.html](https://modula3.github.io/cm3/tutorial/m3/m3_toc.html)

I'm trying to track down this error:

new source -> compiling sandbox.m3
"../sandbox.m3", line 14: actual not assignable to VALUE formal (2.3.2)  (wr)

1 MODULE sandbox EXPORTS Main;
2 
3
4 FROM IO IMPORT Put, PutInt;
5 <*FATAL ANY*>
6 
7 CONST MaxPlaces = 11;
14    PutInt(FIRST(INTEGER), MaxPlaces);Put("\n");

**Comment by thielema** [comment]
On Sun, 2 Jan 2022, Duke Normandin wrote:

> I'm trying to track down this error:
> 
> new source -> compiling sandbox.m3
> "../sandbox.m3", line 14: actual not assignable to VALUE formal (2.3.2) (wr)

It says, that you try to assign the constant MaxPlaces to the second 
parameter of PutInt which has the name "wr".

**Comment by dukester** [comment]
On Sun, 02 Jan 2022 17:45:02 -0800
thielema ***@***.***> wrote:

> On Sun, 2 Jan 2022, Duke Normandin wrote:
>
> > I'm trying to track down this error:
> >
> > new source -> compiling sandbox.m3
> > "../sandbox.m3", line 14: actual not assignable to VALUE formal
> > (2.3.2) (wr)
>
> It says, that you try to assign the constant MaxPlaces to the
> second parameter of PutInt which has the name "wr".

And the solution is?

I - am not trying to do anything of the sort!
I got the code from the Laszlo Boszormenyi/Carsten Weich book:
"Programming in Modula-3". Working through the examples and porting
them to CM3 as needed - if I can.

See: https://www.dropbox.com/s/6jons9npq4qnpaq/M3-min-max-code.jpg?dl=0

--
Duke
** Text only please. Bottom post is best for me **

**Comment by ghost** [comment]
https://modula3.github.io/cm3/reference/complete/html/2_6_13Type_operations.html

The example you're copying uses the "Simple input/output" interface `SIO` which was invented for the book (pg. 50).  The `IO` module has a slightly different interface.

https://modula3.github.io/cm3/help/gen_html/libm3/src/rw/IO.i3.html

**Comment by dukester** [comment]
On Mon, 03 Jan 2022 03:04:47 -0800
Eric Sessoms ***@***.***> wrote:

> https://modula3.github.io/cm3/reference/complete/html/2_6_13Type_operations.html

Thx! Is there an index/glossary - in one location - for everything
in CM3?

> The example you're copying uses the "Simple input/output"
> interface `SIO` which was invented for the book (pg. 50).  The
> `IO` module has a slightly different interface.
>
> https://modula3.github.io/cm3/help/gen_html/libm3/src/rw/IO.i3.html

CM3's IO.PutInt has no "length" field! Nice trick! Modula-3 is not
standardised?

**Comment by RodneyBates** [comment]
On 1/3/22 10:00 AM, Duke Normandin wrote:
> On Mon, 03 Jan 2022 03:04:47 -0800
> Eric Sessoms ***@***.***> wrote:
>
> > https://modula3.github.io/cm3/reference/complete/html/2_6_13Type_operations.html
>
> Thx! Is there an index/glossary - in one location - for everything
> in CM3?
>
> > The example you're copying uses the "Simple input/output"
> > interface `SIO` which was invented for the book (pg. 50). The
> > `IO` module has a slightly different interface.
> >
> > https://modula3.github.io/cm3/help/gen_html/libm3/src/rw/IO.i3.html
>
> CM3's IO.PutInt has no "length" field! Nice trick! Modula-3 is not
> standardised?

SIO is not part of what is standardized, which is Modula-3 itself
and the library code in the required interfaces.  I suppose I would
call the things in "Some useful interfaces" defacto standards.  But
anybody can write their own library code and publish it, like SIO.

It would have been better if the book had either named procedures
differently from widely-used libraries or used qualified references,
i.e. IO.PutInt or SIO.PutInt.  I almost always use qualified references
to anything imported.  Saving confusion later seems more important
that saving keystrokes now.
